### ISSUE TITLE
Interactivity:  Expose bounding rectangle(s) for named nodes

### ISSUE THREAD

**Issue by zathras** [zathras/jovial_svg#92]
Spinning this off from #19, and thinking about @Bindslev's use case, I think "interactivity" is orthogonal to having a DOM.

For "interactivity," hopefully it's sufficient to detect if a mouse click lies within the bounding rectangle of a named node.  This is a little crude -- a path that renders a circle would be detected if the user clicked outside the circle, but within the bounding rectangle.  But actually detecting if a click lands on a pixel that's rendered by a path might not even be feasible -- I think a bounding rectangle is the best that could be done.

@Bindslev, would that be sufficient for your use case?

If so, I note that the bounding rect needs to be computed on the device where the SVG is rendered, because of the text node.  When a .si file is created, the ultimate font metrics are unknown.

Anyway, there is `SIRenderable.getBoundary()` that gives the bounding rectangle.  The missing pieces are:
    *  Associating the name of the SVG node with the appropriate SIRenderable
    *  Reversing any transformations from the root to that node, along all paths from the root to that node
 Note that this may result in multiple bounding rectangles, e.g. if the named node is transitively references by a use node.

So:  An API for this might be:

`ScalableImage.getBoundingRectsFor(String name)`

Also, when parsing an SVG, there would need to be a specification of what name(s) to keep information for, with an easy option for "all of them."  Maybe a list of Pattern objects (thus allowing the regular expression ".*").  This would have to be exposed to the command line when building an SI, too.

**Comment by Bindslev** [comment]
So cool to see work is already being put into this feature!

I think that could definitely work out for my use case. At least I imagine I could probably work around any issues that could arrive from it not being too precise when you click, as I make the SVG's myself. For instance, some of the rooms could be shaped as an L, but I imagine I could divide it up into two paths and perform the same action if any of those two are clicked.

One thing that strikes me when reading your considerations here, is that since I'm making a floor plan, I have put the SVG into an interactiveViewer widget that allows the user to pan and zoom around on the SVG.

Would it still be able to recognize where you clicked when you have zoomed and panned the SVG?

**Comment by zathras** [comment]
> One thing that strikes me when reading your considerations here, is that since I'm making a floor plan, I have put the SVG into an interactiveViewer widget that allows the user to pan and zoom around on the SVG.
> 
> Would it still be able to recognize where you clicked when you have zoomed and panned the SVG?

Yes, provided that your interactive viewer widget exposes the pan/zoom information.  It's a "simple" coordinate transform -- you take the x/y coordinates of the click, and apply the pan/zoom factor that's being viewed to get the x/y coordinates in the coordinate system of the SVG.  The widget is doing the same thing to the `Canvas` object it gets passed in when flutter tells it to paint.

I have most of what would be needed in the exposed class `ScalingTransform`; I probably just need to add a constructor that takes an arbitrary translate and scale.  See https://pub.dev/documentation/jovial_svg/latest/jovial_svg/ScalingTransform-class.html

**Comment by zathras** [comment]
Oh, come to think of it, there's an easier way:  Put the widget that gets the mouse events _inside_ the interactive viewer.  That is, make the widget that holds the `ScalableImage` be the immediate child of the widget that takes the mouse presses.

**Comment by zathras** [comment]
Note to self:  For `ScalableImageWidget`, the `CustomPainter` is of course a `RenderObjectWidget`.  That means I can safely get a `RenderBox` to get me the widget's size like this:
```
  late final GlobalKey _key = GlobalKey();
  ...
  Widget build(BuildContext context) {
     final renderBox = _key.currentContext?.findRenderObject();
     if (renderBox is RenderBox) {
        final Size size = renderBox.size;
        .... do stuff with it
    }
  }
```

Where the key is passed in the CustomPainter constructor.

**Comment by zathras** [comment]
Hi @Bindslev,  I think this is about done.  Let me know if you need a release candidate on pub.dev.

Check out the documentation of `ExportedIDLookup`.  The demo program uses this, too.  If you run the demo and click on "IDs" then click around the `ScalableImage`, you should see a printout of what's under each click on `stdout`.  The second-to-last image is a simple one that exercises some of the less straightforward SVG features for this, like a node with a mask, and a square rotated 45 degrees.

**Comment by Bindslev** [comment]
Awesome stuff @zathras , didn't think you'd pull it off this fast, so I'm very eager to try it out! I'm in Scotland over the easter but when I get back home I'll make sure to test it out and let you know if I have any questions :)

**Comment by Bindslev** [comment]
Hey @zathras

I had the opportunity to test out **exported_ids.si** and my example SVG in your demo project and it worked out of the box. So awesome! Many thanks for implementing this and so swiftly as well. It would be great if you could create a release candidate on pub.dev.

Also, sorry for my ignorance but I was thinking whether it is possible to manipulate the SVG as also mentioned in the old issue? For instance, add a _fill_ or overlay of some sort on the item that was clicked? - I don't actually _need_ this behavior but it would be a cool little design detail when someone clicks a room on my floorplan. As I understand from the other ticket, manipulating the SVG like that may be out of scope since it is done using JavaScript? - Even then it would be very interesting to hear your thoughts about the best way to accomplish this, if possible - also for future readers!

Eitherway, that was just an idea and I consider this issue resolved as it even works better than I had expected so very well done.

**Comment by zathras** [comment]
@Bindslev see comments over at #19

**Comment by zathras** [comment]
Released in 1.1.21
